server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java

```diff
@@ -290,128 +307,119 @@ private void processSegmentChangeRequest()
         // We do not create the existence watcher first, because then it will fire when we create the
         // node and we'll have the same race when trying to refresh that watcher.
         curator.create().withMode(CreateMode.EPHEMERAL).forPath(path, noopPayload);
-
-        entryRemoved(path);
+        entryRemoved(segmentHolder, path);
       }
     }
+    catch (KeeperException.NodeExistsException ne) {
+      // This is expected when historicals haven't yet picked up processing this segment and coordinator
+      // tries reassigning it to the same node.
+      log.warn(ne, "ZK node already exists because segment change request hasn't yet been processed");
+      failAssign(segmentHolder);
+    }
     catch (Exception e) {
-      failAssign(e);
+      failAssign(segmentHolder, e);
     }
   }
 
-  private void actionCompleted()
+  private void actionCompleted(SegmentHolder segmentHolder)
   {
-    if (currentlyProcessing != null) {
-      switch (currentlyProcessing.getType()) {
-        case LOAD:
-          segmentsToLoad.remove(currentlyProcessing.getSegment());
-          queuedSize.addAndGet(-currentlyProcessing.getSegmentSize());
-          break;
-        case DROP:
-          segmentsToDrop.remove(currentlyProcessing.getSegment());
-          break;
-        default:
-          throw new UnsupportedOperationException();
-      }
-
-      final List<LoadPeonCallback> callbacks = currentlyProcessing.getCallbacks();
-      currentlyProcessing = null;
-      callBackExecutor.execute(
-          () -> executeCallbacks(callbacks)
-      );
+    switch (segmentHolder.getType()) {
+      case LOAD:
+        segmentsToLoad.remove(segmentHolder.getSegment());
+        queuedSize.addAndGet(-segmentHolder.getSegmentSize());
+        break;
+      case DROP:
+        segmentsToDrop.remove(segmentHolder.getSegment());
+        break;
+      default:
+        throw new UnsupportedOperationException();
     }
+
+    callBackExecutor.execute(
+        () -> executeCallbacks(segmentHolder)
+    );
   }
 
+
   @Override
   public void start()
   {
-    ScheduledExecutors.scheduleAtFixedRate(
-        processingExecutor,
-        config.getLoadQueuePeonRepeatDelay(),
-        config.getLoadQueuePeonRepeatDelay(),
-        () -> {
-          processSegmentChangeRequest();
-
-          if (stopped) {
-            return ScheduledExecutors.Signal.STOP;
-          } else {
-            return ScheduledExecutors.Signal.REPEAT;
-          }
-        }
-    );
+    for (int i = 0; i < numProcessingQueues; i++) {
+      processingExecutor.scheduleAtFixedRate(
+          new SegmentChangeProcessor(segmentProcessingQueues[i]),
+          0,
+          config.getCuratorCreateZkNodesRepeatDelay().getMillis(),
```

Comment:
Every thread gets its own queue of requests. This helps reduce contention between threads trying to call drainTo(Collection). 

The idea behind repeated runs was to introduce batching so that there is not a flood of zookeeper nodes. Further, it gives historicals some time to service the segment load/drop requests. Without the batching and delay, it is possible that historicals will never get to complete the process segment request causing coordinator to repeatedly delete and create zk nodes.

Reply:
Just wonder how high requests rate we are discussing, 1mln/s?
I'm sure a single queue will manage it. Threads don't need to call *drainTo()* as they can perfectly consume calling *queue.take()*.

Reply:
Just wonder how high requests rate we are discussing, 1mln/s? The Coordinator balancing process will choke to halt from such amount long before the discussed queue will show a mere slight latency. 
I'm sure a single queue will manage the current request rate of Coordinator, besides threads don't need to call *drainTo()* as they can perfectly consume calling *queue.take()*.

Reply:
1mn/second or higher is likely. Coordinator polls for the segment files from the RDS and then proceeds to call load on them in a loop. For a largish cluster, with lots of data and datasources, it is possible to have more than a million segments. 

According to javadocs of BlockingQueue, calling drainTo() is likely more efficient than continuously polling the queue which I think would extend to continuously calling take(). 
```
/**
     * Removes all available elements from this queue and adds them
     * to the given collection.  This operation may be more
     * efficient than repeatedly polling this queue.  
```
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#drainTo-java.util.Collection-

Further, it makes the code simpler since one call can go and fetch the batch worth of records instead of calling take() in a loop.

Reply:
I'm telling about a historical node, did you mean 1 mln/s per historical node?

Reply:
CuratorLoadQueuePeon runs on the coordinator.

Reply:
I'm telling about a historical node, did you mean 1 mln/s per historical node (peon)?

Reply:
Coordinator can read 1mln segments, but it probably does it within minutes (not seconds),  moreover for each segment Coordinator should determine the location (which is the most time-consuming part), then Coordinator spreads those segments across a large pool of peons, overall definitely not 1mln/s. I don't like to complicate peon logic with those queues considering it doesn't give a noticeable performance boost.

Reply:
Fair enough. I have changed it to use a single queue.

Reply:
For smaller clusters it may workout to have just one thread create these zk nodes. For larger ones, we probably need to parallelize the requests. It's configurable.

Reply:
I took a closer look at the code and I realized that I was making the wrong assumption that there is only one peon running on the coordinator. Instead, every historical node gets its own load queue peon on the coordinator. Considering this, I don't think we need to run multiple tasks per peon for creating zk nodes. Just having one task should be enough. Made this change in the latest commit.